// @headlessui-float/react/dist/headlessui-float.mjs
var Wt = Object.defineProperty;
var Yt = (t, a, o) => a in t ? Wt(t, a, { enumerable: !0, configurable: !0, writable: !0, value: o }) : t[a] = o;
var Te = (t, a, o) => (Yt(t, typeof a != "symbol" ? a + "" : a, o), o);
import se, { useEffect as P, useLayoutEffect as Lt, useState as Q, useMemo as q, useRef as z, createContext as Pe, Fragment as j, forwardRef as Mt, isValidElement as tt, useCallback as Nt, useContext as Oe } from "react";
import { Portal as Ut, Transition as Ke } from "@headlessui/react";
import { arrow as Vt, useFloating as Bt } from "@floating-ui/react";
import { offset as zt, flip as Ht, shift as qt, autoPlacement as Xt, hide as Jt, autoUpdate as Gt } from "@floating-ui/dom";
var Se = { exports: {} }, K = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Ze;
function Kt() {
  if (Ze)
    return K;
  Ze = 1;
  var t = se, a = Symbol.for("react.element"), o = Symbol.for("react.fragment"), r = Object.prototype.hasOwnProperty, u = t.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, h = { key: !0, ref: !0, __self: !0, __source: !0 };
  function l(y, d, b) {
    var f, s = {}, C = null, m = null;
    b !== void 0 && (C = "" + b), d.key !== void 0 && (C = "" + d.key), d.ref !== void 0 && (m = d.ref);
    for (f in d)
      r.call(d, f) && !h.hasOwnProperty(f) && (s[f] = d[f]);
    if (y && y.defaultProps)
      for (f in d = y.defaultProps, d)
        s[f] === void 0 && (s[f] = d[f]);
    return { $$typeof: a, type: y, key: C, ref: m, props: s, _owner: u.current };
  }
  return K.Fragment = o, K.jsx = l, K.jsxs = l, K;
}
var Z = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Qe;
function Zt() {
  return Qe || (Qe = 1, process.env.NODE_ENV !== "production" && function() {
    var t = se, a = Symbol.for("react.element"), o = Symbol.for("react.portal"), r = Symbol.for("react.fragment"), u = Symbol.for("react.strict_mode"), h = Symbol.for("react.profiler"), l = Symbol.for("react.provider"), y = Symbol.for("react.context"), d = Symbol.for("react.forward_ref"), b = Symbol.for("react.suspense"), f = Symbol.for("react.suspense_list"), s = Symbol.for("react.memo"), C = Symbol.for("react.lazy"), m = Symbol.for("react.offscreen"), x = Symbol.iterator, O = "@@iterator";
    function W(e) {
      if (e === null || typeof e != "object")
        return null;
      var n = x && e[x] || e[O];
      return typeof n == "function" ? n : null;
    }
    var A = t.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function R(e) {
      {
        for (var n = arguments.length, i = new Array(n > 1 ? n - 1 : 0), c = 1; c < n; c++)
          i[c - 1] = arguments[c];
        Y("error", e, i);
      }
    }
    function Y(e, n, i) {
      {
        var c = A.ReactDebugCurrentFrame, E = c.getStackAddendum();
        E !== "" && (n += "%s", i = i.concat([E]));
        var w = i.map(function(g) {
          return String(g);
        });
        w.unshift("Warning: " + n), Function.prototype.apply.call(console[e], console, w);
      }
    }
    var F = !1, M = !1, N = !1, X = !1, J = !1, ce;
    ce = Symbol.for("react.module.reference");
    function Ie(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === r || e === h || J || e === u || e === b || e === f || X || e === m || F || M || N || typeof e == "object" && e !== null && (e.$$typeof === C || e.$$typeof === s || e.$$typeof === l || e.$$typeof === y || e.$$typeof === d || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === ce || e.getModuleId !== void 0));
    }
    function De(e, n, i) {
      var c = e.displayName;
      if (c)
        return c;
      var E = n.displayName || n.name || "";
      return E !== "" ? i + "(" + E + ")" : i;
    }
    function fe(e) {
      return e.displayName || "Context";
    }
    function $(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && R("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case r:
          return "Fragment";
        case o:
          return "Portal";
        case h:
          return "Profiler";
        case u:
          return "StrictMode";
        case b:
          return "Suspense";
        case f:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case y:
            var n = e;
            return fe(n) + ".Consumer";
          case l:
            var i = e;
            return fe(i._context) + ".Provider";
          case d:
            return De(e, e.render, "ForwardRef");
          case s:
            var c = e.displayName || null;
            return c !== null ? c : $(e.type) || "Memo";
          case C: {
            var E = e, w = E._payload, g = E._init;
            try {
              return $(g(w));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var L = Object.assign, U = 0, le, de, ve, he, me, ge, ye;
    function te() {
    }
    te.__reactDisabledLog = !0;
    function dt() {
      {
        if (U === 0) {
          le = console.log, de = console.info, ve = console.warn, he = console.error, me = console.group, ge = console.groupCollapsed, ye = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: te,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        U++;
      }
    }
    function vt() {
      {
        if (U--, U === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: L({}, e, {
              value: le
            }),
            info: L({}, e, {
              value: de
            }),
            warn: L({}, e, {
              value: ve
            }),
            error: L({}, e, {
              value: he
            }),
            group: L({}, e, {
              value: me
            }),
            groupCollapsed: L({}, e, {
              value: ge
            }),
            groupEnd: L({}, e, {
              value: ye
            })
          });
        }
        U < 0 && R("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var be = A.ReactCurrentDispatcher, Ee;
    function re(e, n, i) {
      {
        if (Ee === void 0)
          try {
            throw Error();
          } catch (E) {
            var c = E.stack.trim().match(/\n( *(at )?)/);
            Ee = c && c[1] || "";
          }
        return `
` + Ee + e;
      }
    }
    var pe = !1, ne;
    {
      var ht = typeof WeakMap == "function" ? WeakMap : Map;
      ne = new ht();
    }
    function We(e, n) {
      if (!e || pe)
        return "";
      {
        var i = ne.get(e);
        if (i !== void 0)
          return i;
      }
      var c;
      pe = !0;
      var E = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var w;
      w = be.current, be.current = null, dt();
      try {
        if (n) {
          var g = function() {
            throw Error();
          };
          if (Object.defineProperty(g.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(g, []);
            } catch (D) {
              c = D;
            }
            Reflect.construct(e, [], g);
          } else {
            try {
              g.call();
            } catch (D) {
              c = D;
            }
            e.call(g.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (D) {
            c = D;
          }
          e();
        }
      } catch (D) {
        if (D && c && typeof D.stack == "string") {
          for (var v = D.stack.split(`
`), S = c.stack.split(`
`), _ = v.length - 1, T = S.length - 1; _ >= 1 && T >= 0 && v[_] !== S[T]; )
            T--;
          for (; _ >= 1 && T >= 0; _--, T--)
            if (v[_] !== S[T]) {
              if (_ !== 1 || T !== 1)
                do
                  if (_--, T--, T < 0 || v[_] !== S[T]) {
                    var k = `
` + v[_].replace(" at new ", " at ");
                    return e.displayName && k.includes("<anonymous>") && (k = k.replace("<anonymous>", e.displayName)), typeof e == "function" && ne.set(e, k), k;
                  }
                while (_ >= 1 && T >= 0);
              break;
            }
        }
      } finally {
        pe = !1, be.current = w, vt(), Error.prepareStackTrace = E;
      }
      var B = e ? e.displayName || e.name : "", Ge = B ? re(B) : "";
      return typeof e == "function" && ne.set(e, Ge), Ge;
    }
    function mt(e, n, i) {
      return We(e, !1);
    }
    function gt(e) {
      var n = e.prototype;
      return !!(n && n.isReactComponent);
    }
    function oe(e, n, i) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return We(e, gt(e));
      if (typeof e == "string")
        return re(e);
      switch (e) {
        case b:
          return re("Suspense");
        case f:
          return re("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case d:
            return mt(e.render);
          case s:
            return oe(e.type, n, i);
          case C: {
            var c = e, E = c._payload, w = c._init;
            try {
              return oe(w(E), n, i);
            } catch {
            }
          }
        }
      return "";
    }
    var ae = Object.prototype.hasOwnProperty, Ye = {}, Le = A.ReactDebugCurrentFrame;
    function ie(e) {
      if (e) {
        var n = e._owner, i = oe(e.type, e._source, n ? n.type : null);
        Le.setExtraStackFrame(i);
      } else
        Le.setExtraStackFrame(null);
    }
    function yt(e, n, i, c, E) {
      {
        var w = Function.call.bind(ae);
        for (var g in e)
          if (w(e, g)) {
            var v = void 0;
            try {
              if (typeof e[g] != "function") {
                var S = Error((c || "React class") + ": " + i + " type `" + g + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[g] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw S.name = "Invariant Violation", S;
              }
              v = e[g](n, g, c, i, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (_) {
              v = _;
            }
            v && !(v instanceof Error) && (ie(E), R("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", c || "React class", i, g, typeof v), ie(null)), v instanceof Error && !(v.message in Ye) && (Ye[v.message] = !0, ie(E), R("Failed %s type: %s", i, v.message), ie(null));
          }
      }
    }
    var bt = Array.isArray;
    function we(e) {
      return bt(e);
    }
    function Et(e) {
      {
        var n = typeof Symbol == "function" && Symbol.toStringTag, i = n && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return i;
      }
    }
    function pt(e) {
      try {
        return Me(e), !1;
      } catch {
        return !0;
      }
    }
    function Me(e) {
      return "" + e;
    }
    function Ne(e) {
      if (pt(e))
        return R("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", Et(e)), Me(e);
    }
    var G = A.ReactCurrentOwner, wt = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, Ue, Ve, xe;
    xe = {};
    function xt(e) {
      if (ae.call(e, "ref")) {
        var n = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (n && n.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function Rt(e) {
      if (ae.call(e, "key")) {
        var n = Object.getOwnPropertyDescriptor(e, "key").get;
        if (n && n.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function Ct(e, n) {
      if (typeof e.ref == "string" && G.current && n && G.current.stateNode !== n) {
        var i = $(G.current.type);
        xe[i] || (R('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', $(G.current.type), e.ref), xe[i] = !0);
      }
    }
    function _t(e, n) {
      {
        var i = function() {
          Ue || (Ue = !0, R("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", n));
        };
        i.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: i,
          configurable: !0
        });
      }
    }
    function Tt(e, n) {
      {
        var i = function() {
          Ve || (Ve = !0, R("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", n));
        };
        i.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: i,
          configurable: !0
        });
      }
    }
    var jt = function(e, n, i, c, E, w, g) {
      var v = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: a,
        // Built-in properties that belong on the element
        type: e,
        key: n,
        ref: i,
        props: g,
        // Record the component responsible for creating this element.
        _owner: w
      };
      return v._store = {}, Object.defineProperty(v._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(v, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: c
      }), Object.defineProperty(v, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: E
      }), Object.freeze && (Object.freeze(v.props), Object.freeze(v)), v;
    };
    function St(e, n, i, c, E) {
      {
        var w, g = {}, v = null, S = null;
        i !== void 0 && (Ne(i), v = "" + i), Rt(n) && (Ne(n.key), v = "" + n.key), xt(n) && (S = n.ref, Ct(n, E));
        for (w in n)
          ae.call(n, w) && !wt.hasOwnProperty(w) && (g[w] = n[w]);
        if (e && e.defaultProps) {
          var _ = e.defaultProps;
          for (w in _)
            g[w] === void 0 && (g[w] = _[w]);
        }
        if (v || S) {
          var T = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          v && _t(g, T), S && Tt(g, T);
        }
        return jt(e, v, S, E, c, G.current, g);
      }
    }
    var Re = A.ReactCurrentOwner, Be = A.ReactDebugCurrentFrame;
    function V(e) {
      if (e) {
        var n = e._owner, i = oe(e.type, e._source, n ? n.type : null);
        Be.setExtraStackFrame(i);
      } else
        Be.setExtraStackFrame(null);
    }
    var Ce;
    Ce = !1;
    function _e(e) {
      return typeof e == "object" && e !== null && e.$$typeof === a;
    }
    function ze() {
      {
        if (Re.current) {
          var e = $(Re.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function Pt(e) {
      {
        if (e !== void 0) {
          var n = e.fileName.replace(/^.*[\\\/]/, ""), i = e.lineNumber;
          return `

Check your code at ` + n + ":" + i + ".";
        }
        return "";
      }
    }
    var He = {};
    function Ot(e) {
      {
        var n = ze();
        if (!n) {
          var i = typeof e == "string" ? e : e.displayName || e.name;
          i && (n = `

Check the top-level render call using <` + i + ">.");
        }
        return n;
      }
    }
    function qe(e, n) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var i = Ot(n);
        if (He[i])
          return;
        He[i] = !0;
        var c = "";
        e && e._owner && e._owner !== Re.current && (c = " It was passed a child from " + $(e._owner.type) + "."), V(e), R('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', i, c), V(null);
      }
    }
    function Xe(e, n) {
      {
        if (typeof e != "object")
          return;
        if (we(e))
          for (var i = 0; i < e.length; i++) {
            var c = e[i];
            _e(c) && qe(c, n);
          }
        else if (_e(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var E = W(e);
          if (typeof E == "function" && E !== e.entries)
            for (var w = E.call(e), g; !(g = w.next()).done; )
              _e(g.value) && qe(g.value, n);
        }
      }
    }
    function Ft(e) {
      {
        var n = e.type;
        if (n == null || typeof n == "string")
          return;
        var i;
        if (typeof n == "function")
          i = n.propTypes;
        else if (typeof n == "object" && (n.$$typeof === d || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        n.$$typeof === s))
          i = n.propTypes;
        else
          return;
        if (i) {
          var c = $(n);
          yt(i, e.props, "prop", c, e);
        } else if (n.PropTypes !== void 0 && !Ce) {
          Ce = !0;
          var E = $(n);
          R("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", E || "Unknown");
        }
        typeof n.getDefaultProps == "function" && !n.getDefaultProps.isReactClassApproved && R("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function kt(e) {
      {
        for (var n = Object.keys(e.props), i = 0; i < n.length; i++) {
          var c = n[i];
          if (c !== "children" && c !== "key") {
            V(e), R("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", c), V(null);
            break;
          }
        }
        e.ref !== null && (V(e), R("Invalid attribute `ref` supplied to `React.Fragment`."), V(null));
      }
    }
    function Je(e, n, i, c, E, w) {
      {
        var g = Ie(e);
        if (!g) {
          var v = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (v += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var S = Pt(E);
          S ? v += S : v += ze();
          var _;
          e === null ? _ = "null" : we(e) ? _ = "array" : e !== void 0 && e.$$typeof === a ? (_ = "<" + ($(e.type) || "Unknown") + " />", v = " Did you accidentally export a JSX literal instead of a component?") : _ = typeof e, R("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", _, v);
        }
        var T = St(e, n, i, E, w);
        if (T == null)
          return T;
        if (g) {
          var k = n.children;
          if (k !== void 0)
            if (c)
              if (we(k)) {
                for (var B = 0; B < k.length; B++)
                  Xe(k[B], e);
                Object.freeze && Object.freeze(k);
              } else
                R("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              Xe(k, e);
        }
        return e === r ? kt(T) : Ft(T), T;
      }
    }
    function At(e, n, i) {
      return Je(e, n, i, !0);
    }
    function $t(e, n, i) {
      return Je(e, n, i, !1);
    }
    var It = $t, Dt = At;
    Z.Fragment = r, Z.jsx = It, Z.jsxs = Dt;
  }()), Z;
}
process.env.NODE_ENV === "production" ? Se.exports = Kt() : Se.exports = Zt();
var p = Se.exports;
class Qt {
  constructor() {
    Te(this, "current", this.detect());
    Te(this, "currentId", 0);
  }
  set(a) {
    this.current !== a && (this.currentId = 0, this.current = a);
  }
  reset() {
    this.set(this.detect());
  }
  nextId() {
    return ++this.currentId;
  }
  get isServer() {
    return this.current === "server";
  }
  get isClient() {
    return this.current === "client";
  }
  detect() {
    return typeof window > "u" || typeof document > "u" ? "server" : "client";
  }
}
const H = new Qt(), rt = (t, a) => {
  H.isServer ? P(t, a) : Lt(t, a);
}, je = { serverHandoffComplete: !1 };
function er() {
  const [t, a] = Q(je.serverHandoffComplete);
  return P(() => {
    t !== !0 && a(!0);
  }, [t]), P(() => {
    je.serverHandoffComplete === !1 && (je.serverHandoffComplete = !0);
  }, []), t;
}
const tr = (
  // Prefer React's `useId` if it's available.
  se.useId ?? function() {
    const a = er(), [o, r] = se.useState(a ? () => H.nextId() : null);
    return rt(() => {
      o === null && r(H.nextId());
    }, [o]), o != null ? `${o}` : void 0;
  }
);
function rr(t, a, o, r) {
  P(() => {
    const u = [];
    (typeof r.offset == "number" || typeof r.offset == "object" || typeof r.offset == "function") && u.push(zt(r.offset)), (r.flip === !0 || typeof r.flip == "number" || typeof r.flip == "object") && u.push(Ht({
      padding: typeof r.flip == "number" ? r.flip : void 0,
      ...typeof r.flip == "object" ? r.flip : {}
    })), (r.shift === !0 || typeof r.shift == "number" || typeof r.shift == "object") && u.push(qt({
      padding: typeof r.shift == "number" ? r.shift : void 0,
      ...typeof r.shift == "object" ? r.shift : {}
    })), (r.autoPlacement === !0 || typeof r.autoPlacement == "object") && u.push(Xt(
      typeof r.autoPlacement == "object" ? r.autoPlacement : void 0
    )), (r.arrow === !0 || typeof r.arrow == "number") && u.push(Vt({
      element: o,
      padding: r.arrow === !0 ? 0 : r.arrow
    })), u.push(...typeof r.middleware == "function" ? r.middleware({
      referenceEl: a.reference,
      floatingEl: a.floating
    }) : r.middleware || []), (r.hide === !0 || typeof r.hide == "object") && u.push(Jt(
      typeof r.hide == "object" ? r.hide : void 0
    )), t(u);
  }, [
    r.offset,
    r.shift,
    r.flip,
    r.arrow,
    r.autoPlacement,
    r.hide,
    r.middleware
  ]);
}
function nr(t, a, o) {
  P(() => {
    if (t && H.isClient && typeof ResizeObserver < "u" && a.current && a.current instanceof Element) {
      const r = new ResizeObserver(([u]) => {
        const h = u.borderBoxSize.reduce((l, { inlineSize: y }) => l + y, 0);
        o(h);
      });
      return r.observe(a.current), () => {
        r.disconnect(), o(null);
      };
    }
  }, []);
}
const Er = [
  "origin-bottom",
  "origin-top",
  "origin-right",
  "origin-left",
  "origin-bottom-left",
  "origin-bottom-right",
  "origin-top-left",
  "origin-top-right"
], or = (t) => {
  switch (t) {
    case "top":
      return "origin-bottom";
    case "bottom":
      return "origin-top";
    case "left":
      return "origin-right";
    case "right":
      return "origin-left";
    case "top-start":
    case "right-end":
      return "origin-bottom-left";
    case "top-end":
    case "left-end":
      return "origin-bottom-right";
    case "right-start":
    case "bottom-start":
      return "origin-top-left";
    case "left-start":
    case "bottom-end":
      return "origin-top-right";
    default:
      return "";
  }
};
function ar(t, a) {
  return q(() => typeof t.originClass == "function" ? t.originClass(a) : typeof t.originClass == "string" ? t.originClass : t.tailwindcssOriginClass ? or(a) : "", [a, t.originClass, t.tailwindcssOriginClass]);
}
function nt(t, a, ...o) {
  if (t in a) {
    const u = a[t];
    return typeof u == "function" ? u(...o) : u;
  }
  const r = new Error(
    `Tried to handle "${t}" but there is no handler defined. Only defined handlers are: ${Object.keys(
      a
    ).map((u) => `"${u}"`).join(", ")}.`
  );
  throw Error.captureStackTrace && Error.captureStackTrace(r, nt), r;
}
function ot(t) {
  return H.isServer ? null : t instanceof Node ? t.ownerDocument : t && Object.prototype.hasOwnProperty.call(t, "current") && t.current instanceof Node ? t.current.ownerDocument : document;
}
const et = [
  "[contentEditable=true]",
  "[tabindex]",
  "a[href]",
  "area[href]",
  "button:not([disabled])",
  "iframe",
  "input:not([disabled])",
  "select:not([disabled])",
  "textarea:not([disabled])"
].map(
  // TODO: Remove this once JSDOM fixes the issue where an element that is
  // "hidden" can be the document.activeElement, because this is not possible
  // in real browsers.
  process.env.NODE_ENV === "test" ? (t) => `${t}:not([tabindex='-1']):not([style*='display: none'])` : (t) => `${t}:not([tabindex='-1'])`
).join(",");
var at = /* @__PURE__ */ ((t) => (t[t.Strict = 0] = "Strict", t[t.Loose = 1] = "Loose", t))(at || {});
function ir(t, a = 0) {
  var o;
  return t === ((o = ot(t)) == null ? void 0 : o.body) ? !1 : nt(a, {
    0() {
      return t.matches(et);
    },
    1() {
      let r = t;
      for (; r !== null; ) {
        if (r.matches(et))
          return !0;
        r = r.parentElement;
      }
      return !1;
    }
  });
}
function ur(t) {
  const a = z(t);
  return rt(() => {
    a.current = t;
  }, [t]), a;
}
function I(t, a, o) {
  const r = ur(a);
  P(() => {
    function u(h) {
      r.current(h);
    }
    return document.addEventListener(t, u, o), () => document.removeEventListener(t, u, o);
  }, [t, o]);
}
function sr(t, a, o = !0) {
  const r = z(!1);
  P(
    process.env.NODE_ENV === "test" ? () => {
      r.current = o;
    } : () => {
      requestAnimationFrame(() => {
        r.current = o;
      });
    },
    [o]
  );
  function u(l, y) {
    if (!r.current || l.defaultPrevented)
      return;
    const d = function f(s) {
      return typeof s == "function" ? f(s()) : Array.isArray(s) || s instanceof Set ? s : [s];
    }(t), b = y(l);
    if (b !== null && b.getRootNode().contains(b)) {
      for (const f of d) {
        if (f === null)
          continue;
        const s = f instanceof HTMLElement ? f : f.current;
        if (s != null && s.contains(b) || l.composed && l.composedPath().includes(s))
          return;
      }
      return (
        // This check alllows us to know whether or not we clicked on a "focusable" element like a
        // button or an input. This is a backwards compatibility check so that you can open a <Menu
        // /> and click on another <Menu /> which should close Menu A and open Menu B. We might
        // revisit that so that you will require 2 clicks instead.
        !ir(b, at.Loose) && // This could be improved, but the `Combobox.Button` adds tabIndex={-1} to make it
        // unfocusable via the keyboard so that tabbing to the next item from the input doesn't
        // first go to the button.
        b.tabIndex !== -1 && l.preventDefault(), a(l, b)
      );
    }
  }
  const h = z(null);
  I("mousedown", (l) => {
    var y, d;
    r.current && (h.current = ((d = (y = l.composedPath) == null ? void 0 : y.call(l)) == null ? void 0 : d[0]) || l.target);
  }, !0), I(
    "click",
    (l) => {
      h.current && (u(l, () => h.current), h.current = null);
    },
    // We will use the `capture` phase so that layers in between with `event.stopPropagation()`
    // don't "cancel" this outside click check. E.g.: A `Menu` inside a `DialogPanel` if the `Menu`
    // is open, and you click outside of it in the `DialogPanel` the `Menu` should close. However,
    // the `DialogPanel` has a `onClick(e) { e.stopPropagation() }` which would cancel this.
    !0
  ), I("blur", (l) => u(
    l,
    () => window.document.activeElement instanceof HTMLIFrameElement ? window.document.activeElement : null
  ), !0);
}
const ue = /* @__PURE__ */ new Map(), Fe = Pe(null);
Fe.displayName = "ReferenceContext";
const ke = Pe(null);
ke.displayName = "FloatingContext";
const ee = Pe(null);
ee.displayName = "ArrowContext";
function it(t) {
  const a = Oe(Fe);
  if (a === null) {
    const o = new Error(`<${t} /> is missing a parent <Float /> component.`);
    throw Error.captureStackTrace && Error.captureStackTrace(o, it), o;
  }
  return a;
}
function ut(t) {
  const a = Oe(ke);
  if (a === null) {
    const o = new Error(`<${t} /> is missing a parent <Float /> component.`);
    throw Error.captureStackTrace && Error.captureStackTrace(o, ut), o;
  }
  return a;
}
function st(t) {
  const a = Oe(ee);
  if (a === null) {
    const o = new Error(`<${t} /> is missing a parent <Float /> component.`);
    throw Error.captureStackTrace && Error.captureStackTrace(o, st), o;
  }
  return a;
}
function ct(t, a, o, r) {
  const { referenceRef: u } = r, h = a;
  if (h.as === j)
    return /* @__PURE__ */ p.jsx(
      t.type,
      {
        ...t.props,
        ...o,
        ref: u
      }
    );
  const l = h.as || "div";
  return /* @__PURE__ */ p.jsx(l, { ...o, ref: u, children: /* @__PURE__ */ p.jsx(t.type, { ...t.props }) });
}
function Ae(t, a, o, r) {
  const { floatingRef: u, props: h, mounted: l, setShow: y, x: d, y: b, placement: f, strategy: s, referenceElWidth: C } = r, m = {
    ...h,
    ...a
  }, x = ar(m, f), O = {
    show: l.current ? m.show : !1,
    enter: `${m.enter || ""} ${x}`,
    enterFrom: `${m.enterFrom || ""}`,
    enterTo: `${m.enterTo || ""}`,
    leave: `${m.leave || ""} ${x}`,
    leaveFrom: `${m.leaveFrom || ""}`,
    leaveTo: `${m.leaveTo || ""}`,
    beforeEnter: () => {
      y(!0);
    },
    afterLeave: () => {
      y(!1);
    }
  }, W = {
    style: {
      // If enable dialog mode, then set `transform` to false.
      ...!m.dialog && (m.transform || m.transform === void 0) ? {
        position: s,
        zIndex: m.zIndex || 9999,
        top: "0px",
        left: "0px",
        right: "auto",
        bottom: "auto",
        transform: `translate(${Math.round(d || 0)}px,${Math.round(b || 0)}px)`
      } : {
        position: s,
        zIndex: m.zIndex || 9999,
        top: `${b || 0}px`,
        left: `${d || 0}px`
      },
      width: m.adaptiveWidth && typeof C == "number" ? `${C}px` : void 0
    }
  };
  function A(F) {
    return m.portal ? /* @__PURE__ */ p.jsx(Ut, { children: F }) : F;
  }
  function R(F) {
    const M = {
      ...W,
      ...o,
      ref: u
    };
    if (m.as === j)
      return /* @__PURE__ */ p.jsx(
        F.type,
        {
          ...F.props,
          ...M
        }
      );
    const N = m.as || "div";
    return /* @__PURE__ */ p.jsx(N, { ...M, children: /* @__PURE__ */ p.jsx(F.type, { ...F.props }) });
  }
  function Y() {
    return H.isServer ? l.current && m.show ? /* @__PURE__ */ p.jsx(t.type, { ...t.props }) : /* @__PURE__ */ p.jsx(j, {}) : m.transitionChild ? /* @__PURE__ */ p.jsx(Ke.Child, { as: j, ...O, children: /* @__PURE__ */ p.jsx(t.type, { ...t.props }) }) : /* @__PURE__ */ p.jsx(Ke, { as: j, ...O, children: /* @__PURE__ */ p.jsx(t.type, { ...t.props }) });
  }
  return A(
    R(
      Y()
    )
  );
}
function ft([t, a], o) {
  var X, J;
  const r = tr(), u = z(!1), [h, l] = Q(), y = z(null), d = q(() => ({
    show: o.onShow || (() => {
    }),
    hide: o.onHide || (() => {
    }),
    update: o.onUpdate || (() => {
    })
  }), [o.onShow, o.onHide, o.onUpdate]), { x: b, y: f, placement: s, strategy: C, update: m, refs: x, middlewareData: O } = Bt({
    placement: o.placement || "bottom-start",
    strategy: o.strategy,
    middleware: h
  }), [W, A] = Q(null);
  P(() => {
    u.current = !0;
  }, []), P(() => {
    t && !ue.get(r) ? (ue.set(r, !0), d.show()) : !t && ue.get(r) && (ue.delete(r), d.hide());
  }, [t]);
  const R = Nt(() => {
    m(), d.update();
  }, [m, d]);
  P(R, [o.placement, o.strategy, h]), rr(l, x, y, o), nr(o.adaptiveWidth, x.reference, A), P(() => {
    if (x.reference.current && x.floating.current && t)
      return o.autoUpdate !== !1 ? Gt(
        x.reference.current,
        x.floating.current,
        R,
        typeof o.autoUpdate == "object" ? o.autoUpdate : void 0
      ) : () => {
      };
  }, [t, R, x]);
  const Y = z(!0);
  P(() => {
    !(x.reference.current instanceof Element) && x.reference.current && x.floating.current && Y.current && (Y.current = !1, R(), window.requestAnimationFrame(() => {
      Y.current = !0, R();
    }));
  }, [x]);
  const F = {
    referenceRef: x.setReference,
    placement: s
  }, M = {
    floatingRef: x.setFloating,
    props: o,
    mounted: u,
    setShow: a,
    x: b,
    y: f,
    placement: s,
    strategy: C,
    referenceElWidth: W
  }, N = {
    arrowRef: y,
    placement: s,
    x: (X = O.arrow) == null ? void 0 : X.x,
    y: (J = O.arrow) == null ? void 0 : J.y
  };
  return { referenceApi: F, floatingApi: M, arrowApi: N, x: b, y: f, placement: s, strategy: C, update: R, refs: x, middlewareData: O };
}
const lt = Mt((t, a) => {
  const [o, r] = Q(t.show ?? !1), {
    referenceApi: u,
    floatingApi: h,
    arrowApi: l,
    placement: y
  } = ft([o, r], t), d = { placement: y }, [b, f] = typeof t.children == "function" ? t.children(d) : t.children;
  if (!tt(b))
    return console.warn("<Float /> is missing a reference and floating element."), /* @__PURE__ */ p.jsx(j, {});
  function s(x) {
    if (t.as === j || !t.as)
      return /* @__PURE__ */ p.jsx(j, { children: x });
    const O = t.as;
    return /* @__PURE__ */ p.jsx(O, { ref: a, className: t.className, children: x });
  }
  if (t.composable || t.dialog)
    return s(
      /* @__PURE__ */ p.jsx(Fe.Provider, { value: u, children: /* @__PURE__ */ p.jsx(ke.Provider, { value: h, children: /* @__PURE__ */ p.jsx(ee.Provider, { value: l, children: typeof t.children == "function" ? t.children(d) : t.children }) }) }, "FloatingNode")
    );
  const C = ct(
    b,
    { as: j },
    { key: "reference-node" },
    u
  ), m = Ae(
    f,
    { as: t.floatingAs || "div" },
    {},
    h
  );
  return s([
    C,
    /* @__PURE__ */ p.jsx(ee.Provider, { value: l, children: m }, "floating-node")
  ]);
});
lt.displayName = "Float";
function cr(t) {
  if (!t.children)
    return /* @__PURE__ */ p.jsx(j, {});
  const a = q(() => {
    const { as: h, children: l, ...y } = t;
    return y;
  }, [t]), o = it("Float.Reference"), { placement: r } = o, u = { placement: r };
  return ct(
    typeof t.children == "function" ? t.children(u) : t.children,
    { ...t, as: t.as || j },
    a,
    o
  );
}
function fr(t) {
  if (!t.children)
    return /* @__PURE__ */ p.jsx(j, {});
  const a = q(() => {
    const { as: h, enter: l, enterFrom: y, enterTo: d, leave: b, leaveFrom: f, leaveTo: s, originClass: C, tailwindcssOriginClass: m, transitionChild: x, children: O, ...W } = t;
    return W;
  }, [t]), o = ut("Float.Content"), { placement: r } = o, u = { placement: r };
  return Ae(
    typeof t.children == "function" ? t.children(u) : t.children,
    { ...t, as: t.as || "div" },
    a,
    o
  );
}
function lr(t) {
  const { arrowRef: a, placement: o, x: r, y: u } = st("Float.Arrow"), h = q(() => {
    const { as: b, offset: f, children: s, ...C } = t;
    return C;
  }, [t]), l = {
    top: "bottom",
    right: "left",
    bottom: "top",
    left: "right"
  }[o.split("-")[0]], y = {
    left: typeof r == "number" ? `${r}px` : void 0,
    top: typeof u == "number" ? `${u}px` : void 0,
    right: void 0,
    bottom: void 0,
    [l]: `${(t.offset ?? 4) * -1}px`,
    ...h.style
  };
  if (t.as === j) {
    const b = { placement: o }, f = typeof t.children == "function" ? t.children(b) : t.children;
    return !f || !tt(f) ? /* @__PURE__ */ p.jsx(j, {}) : /* @__PURE__ */ p.jsx(
      f.type,
      {
        ...f.props,
        ref: a,
        style: y
      }
    );
  }
  const d = t.as || "div";
  return /* @__PURE__ */ p.jsx(
    d,
    {
      ref: a,
      ...h,
      style: y,
      children: t.children
    }
  );
}
function $e({ onInitial: t, children: a, ...o }) {
  const [r, u] = Q(o.show ?? !1), h = q(() => {
    const { as: m, show: x, placement: O, strategy: W, offset: A, shift: R, flip: Y, arrow: F, autoPlacement: M, hide: N, autoUpdate: X, zIndex: J, enter: ce, enterFrom: Ie, enterTo: De, leave: fe, leaveFrom: $, leaveTo: L, originClass: U, tailwindcssOriginClass: le, portal: de, transform: ve, middleware: he, onShow: me, onHide: ge, onUpdate: ye, ...te } = o;
    return te;
  }, [o]), {
    floatingApi: l,
    arrowApi: y,
    placement: d,
    refs: b
  } = ft([r, u], o);
  P(() => {
    u(o.show ?? !1);
  }, [o.show]);
  function f() {
    r && u(!1);
  }
  if (t({ show: r, setShow: u, placement: d, refs: b }), !a)
    return /* @__PURE__ */ p.jsx(j, {});
  const C = Ae(
    typeof a == "function" ? a({ placement: d, close: f }) : a,
    {
      ...o,
      as: o.as || j,
      show: r
    },
    h,
    l
  );
  return /* @__PURE__ */ p.jsx(ee.Provider, { value: y, children: C });
}
function dr(t) {
  function a({ setShow: o, refs: r }) {
    I("contextmenu", (u) => {
      u.preventDefault(), r.setPositionReference({
        getBoundingClientRect() {
          return {
            width: 0,
            height: 0,
            x: u.clientX,
            y: u.clientY,
            top: u.clientY,
            left: u.clientX,
            right: u.clientX,
            bottom: u.clientY
          };
        }
      }), o(!0);
    }), sr(r.floating, () => {
      o(!1);
    });
  }
  return /* @__PURE__ */ p.jsx(
    $e,
    {
      flip: !0,
      ...t,
      show: !1,
      portal: !0,
      onInitial: a
    }
  );
}
function vr({ globalHideCursor: t, ...a }) {
  function o({ setShow: r, refs: u }) {
    function h() {
      r(!0);
    }
    function l() {
      r(!1);
    }
    function y(s) {
      u.setPositionReference({
        getBoundingClientRect() {
          return {
            width: 0,
            height: 0,
            x: s.clientX,
            y: s.clientY,
            top: s.clientY,
            left: s.clientX,
            right: s.clientX,
            bottom: s.clientY
          };
        }
      });
    }
    function d(s) {
      h(), y(s);
    }
    function b(s) {
      h(), y(s.touches[0]);
    }
    const f = ot(u.floating);
    f && (P(() => {
      if ((t || t === void 0) && !f.getElementById("headlesui-float-cursor-style")) {
        const s = f.createElement("style");
        return (f.head || f.getElementsByTagName("head")[0]).appendChild(s), s.id = "headlesui-float-cursor-style", s.appendChild(f.createTextNode([
          "*, *::before, *::after {",
          "  cursor: none !important;",
          "}",
          ".headlesui-float-cursor-root {",
          "  pointer-events: none !important;",
          "}"
        ].join(`
`))), () => {
          var m;
          return (m = f.getElementById("headlesui-float-cursor-style")) == null ? void 0 : m.remove();
        };
      }
    }, [t]), "ontouchstart" in window || navigator.maxTouchPoints > 0 ? (I("touchstart", b), I("touchend", l), I("touchmove", b)) : (I("mouseenter", d), I("mouseleave", l), I("mousemove", d)));
  }
  return /* @__PURE__ */ p.jsx(
    $e,
    {
      ...a,
      portal: !0,
      className: "headlesui-float-cursor-root",
      onInitial: o
    }
  );
}
const pr = Object.assign(lt, { Reference: cr, Content: fr, Arrow: lr, Virtual: $e, ContextMenu: dr, Cursor: vr });
export {
  pr as Float,
  Ae as renderFloatingElement,
  ct as renderReferenceElement,
  or as tailwindcssOriginClassResolver,
  Er as tailwindcssOriginSafelist,
  sr as useOutsideClick
};
